### PR TITLE
MGDAPI-5929 - Use correct ingress router to determine IP for 3scale dns alerts

### PR DESCRIPTION
**Pull request by valerymo**
# Issue link
https://issues.redhat.com/browse/MGDAPI-5929
Use the correct ingress router to determine IP for 3scale dns alerts

# Verification steps
- Check that ***DNSByPass*** alerts are not firing in Custom Domain Rhoam installation
- Regression - check alerts in regular (Not custom domain) Rhoam  installation

- **References**:  
[Configuring custom domain](https://integreatly-operator.readthedocs.io/en/latest/products/custom_domain/)

### CODE REVIEW

File: pkg/products/threescale/reconciler.go [hunk 1]
```diff
@@ -390,30 +390,46 @@ func (r *Reconciler) Reconcile(ctx context.Context, installation *integreatlyv1a
...
 	}
 
-	ingressRouterService, err := customDomain.GetIngressRouterService(ctx, serverClient)
-	if err != nil || len(ingressRouterService.Status.LoadBalancer.Ingress) == 0 {
-		errorMessage := "failed to retrieve ingress router service"
-		r.log.Error(errorMessage, err)
-		phase = integreatlyv1alpha1.PhaseFailed
-		events.HandleError(r.recorder, installation, phase, errorMessage, err)
-		return phase, fmt.Errorf("%s: %v", errorMessage, err)
+	threescaleMasterRoute, err := r.getThreescaleRoute(ctx, serverClient, "system-master", nil)
```
Comment (briangallagher) [inline pkg/products/threescale/reconciler.go:393]: We can not use the ingress-router from the the system-master route for the system provider and system-develoepr. The ingress router may be different on those

File: pkg/products/threescale/reconciler.go [hunk 2]
```diff
@@ -390,30 +390,46 @@ func (r *Reconciler) Reconcile(ctx context.Context, installation *integreatlyv1a
...
 
-	ips, err := customDomain.GetIngressRouterIPs(ingressRouterService.Status.LoadBalancer.Ingress)
-	if err != nil {
-		errorMessage := "failed to retrieve ingress router ips"
-		r.log.Error(errorMessage, err)
-		phase = integreatlyv1alpha1.PhaseFailed
-		events.HandleError(r.recorder, installation, phase, errorMessage, err)
-		return phase, fmt.Errorf("%s: %v", errorMessage, err)
+	listOfIngresServicesNames := []string{"router-default"}
+	if customDomainActive {
```
Comment (briangallagher) [inline pkg/products/threescale/reconciler.go:399]: We should not only do this if custom domain active. There maybe another ingress router even if this is false. Probably unlikely but we should check anyway

File: pkg/products/threescale/reconciler.go [hunk 3]
```diff
@@ -390,30 +390,46 @@ func (r *Reconciler) Reconcile(ctx context.Context, installation *integreatlyv1a
...
+		// get ingres custom domain service (LoadBalancer, router) name from 3scale route
+		if len(threescaleMasterRoute.Status.Ingress) > 1 {
```
Comment (briangallagher) [inline pkg/products/threescale/reconciler.go:401]: We can just take the first item in the array. If there are multiple Ingress that just means the route is on multiple. The first in the list should always work

Comment (valerymo) [reply pkg/products/threescale/reconciler.go:401]: the first item - is default . It's always "router-default". Here we are checking if we have more then one router (len > 1), and in this case we are taking the second one, that is CustomDomain

Comment (valerymo) [reply pkg/products/threescale/reconciler.go:401]: The first in the list - is always working

**Comment by briangallagher** [comment]
I think we should add the logic around [here](https://github.com/valerymo/integreatly-operator/blob/db756b1e77d41fc0bb3e9d0cce1864982751a59d/pkg/products/threescale/reconciler.go#L3342) 

After we get each route, look up the ingress router and then get the IP and use that for the ping.
We should assume that the ingress router could be different for each one of the 3scale routes - this is unlikely but possible.,

**Comment by valerymo** [comment]
> I think we should add the logic around [here](https://github.com/valerymo/integreatly-operator/blob/db756b1e77d41fc0bb3e9d0cce1864982751a59d/pkg/products/threescale/reconciler.go#L3342)
> 
> After we get each route, look up the ingress router and then get the IP and use that for the ping. We should assume that the ingress router could be different for each one of the 3scale routes - this is unlikely but possible.,

OK, I will try change it. Will do on separate commit. Thank you

**Comment by valerymo** [comment]
> I think we should add the logic around [here](https://github.com/valerymo/integreatly-operator/blob/db756b1e77d41fc0bb3e9d0cce1864982751a59d/pkg/products/threescale/reconciler.go#L3342)
> 
> After we get each route, look up the ingress router and then get the IP and use that for the ping. We should assume that the ingress router could be different for each one of the 3scale routes - this is unlikely but possible.,

Done. Logic was moved to ping3scalePortals. It impacted on unittest, but tests as seems to me do nothing almost, and I commented it; as result - small  (-0.06%) reducing of  coverage. Maybe we can accept it?

File: pkg/products/threescale/reconciler.go (hunk 3 above)
Comment (valerymo) [reply pkg/products/threescale/reconciler.go]: Logic was moved to ping3scalePortals

File: pkg/products/threescale/reconciler.go (hunk 2 above)
Comment (valerymo) [reply pkg/products/threescale/reconciler.go]: Logic was changed and moved ping3scalePortals

File: pkg/products/threescale/reconciler.go (hunk 1 above)
Comment (valerymo) [reply pkg/products/threescale/reconciler.go]: Logic was changed and moved to ping3scalePortals.

File: pkg/products/threescale/reconciler.go [hunk 4]
```diff
@@ -3308,50 +3290,58 @@ func (r *Reconciler) findCustomDomainCr(ctx context.Context, serverClient k8scli
...
+			prefix = "3scale-admin."
+		}
+		route, err := r.getThreescaleRoute(ctx, serverClient, threeScaleService, func(r routev1.Route) bool {
+			return strings.HasPrefix(r.Spec.Host, prefix)
+		})
+		if err != nil || route == nil {
+			return integreatlyv1alpha1.PhaseFailed, fmt.Errorf(format, threeScaleService)
+		}
+		ingressRouters := []string{"router-default"}
+		if len(route.Status.Ingress) > 1 {
```
Comment (briangallagher) [inline pkg/products/threescale/reconciler.go:3316]: I don't think we can assume that the first route will be the default one. On the latest OHSS customer issue this is not the case. The 3scale route only has one ingress on it and it's not the default one.

Comment (valerymo) [reply pkg/products/threescale/reconciler.go:3316]: before write it i looked at 3scale, and seems "router-default" is always present .. But I will look how to handle it

File: pkg/products/threescale/reconciler.go [hunk 5]
```diff
@@ -3308,50 +3290,58 @@ func (r *Reconciler) findCustomDomainCr(ctx context.Context, serverClient k8scli
...
 	portals := map[string]metrics.PortalInfo{}
+	ips := make(map[string][]net.IP)
+	metricsLabelMap := make(map[string]string)
+	metricsLabelMap[labelRouteToSystemDeveloper] = metrics.LabelSystemDeveloper
+	metricsLabelMap[labelRouteToSystemProvider] = metrics.LabelSystemProvider
+	metricsLabelMap[labelRouteToSystemMaster] = metrics.LabelSystemMaster
+
+	for _, threeScaleService := range [3]string{labelRouteToSystemDeveloper, labelRouteToSystemProvider, labelRouteToSystemMaster} {
+		var prefix string
+		if threeScaleService == labelRouteToSystemDeveloper {
```
Comment (briangallagher) [inline pkg/products/threescale/reconciler.go:3304]: We need to return to original filtering logic

File: pkg/products/threescale/reconciler_test.go [hunk 6]
```diff
@@ -2948,6 +2964,21 @@ func TestReconciler_ping3scalePortals(t *testing.T) {
 			},
 		},
 	}
+	openshiftIngress := corev1.Service{
```
Comment (austincunningham) [inline pkg/products/threescale/reconciler_test.go:2967]: @valerymo do we need to define the same object twice can we do without one of these.

Comment (austincunningham) [reply pkg/products/threescale/reconciler_test.go:2967]: never mind looks like they are in different functions

File: pkg/products/threescale/reconciler.go (hunk 4 above)
Comment (valerymo) [reply pkg/products/threescale/reconciler.go]: done, as suggested:  https://github.com/integr8ly/integreatly-operator/pull/3338

File: pkg/products/threescale/reconciler_test.go (hunk 6 above)
Comment (valerymo) [reply pkg/products/threescale/reconciler_test.go:2967]: Thank you for help @austincunningham

**Comment by carlkyrillos** [comment]
I reviewed the changed files and verified that the `DnsBypassThreeScaleUI` alerts don't fire when doing a custom domain installation.
/lgtm

**Comment by carlkyrillos** [comment]
If we later want to configure the 3scale routes to only be serviceable by the custom domain IngressController,  this `oc` command can be used to update the default IngressController to filter out the 3scale routes:
```bash
oc patch ingresscontroller default -n openshift-ingress-operator --type=json -p '[{"op":"replace", "path":"/spec/routeSelector/matchExpressions", "value": [{"key": "zync.3scale.net/route-to", "operator": "DoesNotExist"}]}]'
```

CC: @briangallagher

**Comment by austincunningham** [comment]
/lgtm
After reviewing the code changes, since the rebase and I agree the verification stands as there are no changes to the actual functional code. I will approve once the e2e pass

**Comment by valerymo** [comment]
retested on custom domain  (cluster vmogilev01) and regression - for non custom domain (cluster vmogilev02). No DNS alerts

**Pull request merged**